kv/memdb.go

```diff
@@ -0,0 +1,460 @@
+// Copyright 2015 PingCAP, Inc.
+//
+// Copyright 2015 Wenbin Xiao
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package kv
+
+import (
+	"bytes"
+	"context"
+	"sync/atomic"
+
+	"github.com/pingcap/errors"
+)
+
+var (
+	_ MemBuffer     = new(memDB)
+	_ StagingBuffer = sandboxWrapper{}
+)
+
+// KeyFlags are metadata associated with key
+type KeyFlags uint8
+
+const (
+	flagPresumeKeyNotExists KeyFlags = 1 << iota
+
+	flagValidMark KeyFlags = 0x80
+)
+
+// MarkPresumeKeyNotExists marks the existence of the associated key is checked lazily.
+func (m KeyFlags) MarkPresumeKeyNotExists() KeyFlags {
+	return (m | flagPresumeKeyNotExists) | flagValidMark
+}
+
+// UnmarkPresumeKeyNotExists reverts MarkPresumeKeyNotExists.
+func (m KeyFlags) UnmarkPresumeKeyNotExists() KeyFlags {
+	return (m & ^flagPresumeKeyNotExists) | flagValidMark
+}
+
+// HasPresumeKeyNotExists retruns whether the associated key use lazy check.
+func (m KeyFlags) HasPresumeKeyNotExists() bool {
+	return m&flagPresumeKeyNotExists != 0
```

Comment:
Mark, then Unmark, HasPresumeKeyNotExists returns true?

Reply:
Expose this API may easy lead to misuse as `isValid` is not checked.

Reply:
It will be false, https://play.golang.org/p/Uq7mXSyMUdX